plugins/catalog-backend/src/ingestion/LocationReaders.ts

```diff
@@ -0,0 +1,207 @@
+  ) {
+    this.logger = logger;
+    this.processors = processors;
+  }
+
+  async read(location: LocationSpec): Promise<ReadLocationResult> {
+    const output: ReadLocationResult = { entities: [], errors: [] };
+    const initialItem: LocationProcessorResult = {
+      type: 'location',
+      location,
+      optional: false,
+    };
+    await this.handleResultItem(initialItem, 0, output);
+    return output;
+  }
+
+  async handleResultItem(
+    item: LocationProcessorResult,
+    depth: number,
+    output: ReadLocationResult,
+  ): Promise<void> {
+    // Sanity check to break silly expansions / loops
+    if (depth > MAX_DEPTH) {
+      const message = `Max recursion depth ${MAX_DEPTH} reached at ${item.location.type} ${item.location.target}`;
+      this.logger.warn(message);
+      output.errors.push({
+        location: item.location,
+        error: new Error(message),
+      });
+      return;
+    }
+
+    if (item.type === 'location') {
+      this.logger.debug(
+        `Reading location ${item.location.type} ${item.location.target} optional=${item.optional}`,
+      );
+      await this.runAll(
+        'fetch',
+        processor => processor.readLocation?.(item.location, item.optional),
+        emitted => this.handleResultItem(emitted, depth + 1, output),
+        item.location,
+        true,
+        true,
+      );
+    } else if (item.type === 'data') {
+      this.logger.debug(
+        `Parsing data from location ${item.location.type} ${item.location.target} (${item.data.byteLength} bytes)`,
+      );
+      await this.runAll(
+        'parse',
+        processor => processor.parseData?.(item.data, item.location),
+        emitted => this.handleResultItem(emitted, depth + 1, output),
+        item.location,
+        true,
+        true,
+      );
+    } else if (item.type === 'error') {
+      this.logger.debug(
+        `Encountered error at location ${item.location.type} ${item.location.target}, ${item.error}`,
+      );
+      await this.runAll(
+        'process error',
+        processor => processor.handleError?.(item.error, item.location),
+        emitted => this.handleResultItem(emitted, depth + 1, output),
+        item.location,
+        false,
+        false,
+      );
+      output.errors.push({
+        location: item.location,
+        error: item.error,
+      });
+    } else if (item.type === 'entity') {
+      this.logger.debug(
+        `Got entity at location ${item.location.type} ${item.location.target}, ${item.entity.apiVersion} ${item.entity.kind}`,
+      );
+      const current = { entity: item.entity, location: item.location };
+      await this.runAll(
+        'process entity',
+        processor =>
+          processor.processEntity?.(current.entity, current.location),
+        async emitted => {
+          if (emitted.type === 'entity') {
+            current.entity = emitted.entity;
+            current.location = emitted.location;
```

Comment:
This'll override previously emitted entities right? Which I don't think is what we want

Reply:
No it's right. We effectively want `emit(f1(f2(f3(f4(originalEntity))))`

Reply:
This is not a simple function composition though? What if `f2` emits 2 entities?

Reply:
Yeah that's a point. Emitting more than one would be unexpected here, but unfortunately permitted by the api.

Reply:
I could change the signature for processEntity to return `Promise<Entity>`

Reply:
changed this to do explicit chaining in handleEntity